### ISSUE TITLE
unit_arg: can't ignore lint and can't disable it

### ISSUE THREAD

**Issue by xobs** [rust-lang/rust-clippy#5728]
Clippy version: `clippy 0.0.212 (204bb9b5 2020-03-17)` (the most recent one from rustup)

I have an operating system that runs on multiple platforms. Each process has a state that's an enum. When a process first gets set up, it has an "initial context" that's platform dependent.

On many platforms, this context contains things like the initial program counter and stack pointer. However, on some platforms is empty, so it's defined as `pub type ContextInit = ();`

Clippy is warning me about this:

```rust
            if idx == 0 {
                entry.state = ProcessState::Running(0);
            } else {
                entry.state = ProcessState::Setup(init_context);
            }
```

It's saying I'm passing `()` to a function:

```
passing a unit value to a function
`#[warn(clippy::unit_arg)]` on by default
for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unit_arg
```

There are multiple issues here:

* It's not a function, it's an enum
* It's only true on one platform, so I would like to ignore this lint
* This lint cannot be disabled:

```rust
            if idx == 0 {
                entry.state = ProcessState::Running(0);
            } else {
                #[allow(clippy::unit_arg)]
                entry.state = ProcessState::Setup(init_context);
            }
```

results in:

```sh
> cargo run
   Compiling kernel v0.1.0 (D:\Code\Xous\core\kernel)
error[E0658]: attributes on expressions are experimental
   --> kernel\src\services.rs:248:17
    |
248 |                 #[allow(clippy::unit_arg)]
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: see issue #15701 <https://github.com/rust-lang/rust/issues/15701> for more information
...
```

**Comment by ebroto** [comment]
>It's not a function, it's an enum

I think it's referring to the constructor of the enum variant

>This lint cannot be disabled

To workaround attributes on expressions being experimental, as suggested in the linked issue you could wrap the assignment in a block and attach the attribute to it
```rust
        #[allow(clippy::unit_arg)]
        { entry.state = ProcessState::Setup(init_context); }
```
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=0232ad196faec2f8265a6e2798b45c59))

Other options are attaching the attribute to the `if`, or maybe using an empty struct instead of the unit type.

**Comment by xobs** [comment]
If I wrap it in a block and attach an attribute to that, it presents a new lint:

```
statement can be reduced
`#[warn(clippy::unnecessary_operation)]` on by default
for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_operation
```

This similarly produces the same warning:
```
#[allow(clippy::unit_arg,clippy::unnecessary_operation)]
{entry.state = ProcessState::Setup(init_context)};
```

As does this:

```
#[allow(clippy::unit_arg)]
#[allow(clippy::unnecessary_operation)]
{entry.state = ProcessState::Setup(init_context)};
```

Attaching it to the `if` statement appears to work, except it now produces an error in `rust-analyzer`:

```
Syntax Error: attributes are not allowed on IF_EXPR
```

**Comment by ebroto** [comment]
>If I wrap it in a block and attach an attribute to that, it presents a new lint

The semicolon should be after the constructor call, not after the block

>Attaching it to the if statement appears to work, except it now produces an error in rust-analyzer

Attributes on `if`s are quite new, maybe they are not yet supported in rust analyzer

**Comment by flip1995** [comment]
IIUC, @ebroto already addressed every concern you have listed.

> * It's not a function, it's an enum

`enum` constructors are considered functions in Rust

> * It's only true on one platform, so I would like to ignore this lint
> * This lint cannot be disabled:

You can ignore this lint and the behavior is as expected. Clippy can't change the fact that "`attributes on expressions are experimental`"

> Attaching it to the if statement appears to work, except it now produces an error in rust-analyzer

This is a thing between RA and rustc and there's nothing Clippy can do here.

---

Sorry that the answer here is "Not Clippy's problem", but there's really nothing we can change in Clippy to address attribute handling.

If you have any other concerns about this lint, let me know and I'll reopen the issue.
